lib/models/addon.js

```diff
@@ -371,16 +391,76 @@ let addonProto = {
   },
 
   /**
-    Discovers all child addons of this addon and stores their names and
-    package.json contents in this.addonPackages as key-value pairs
-
-    @private
-    @method discoverAddons
+   * Discovers all child addons of this addon and an AddonInfo about
+   * each addon in this.addonPackages (keyed on addon name).
+   *
+   * Child addons include those from 'dependencies' (not devDependencies)
+   * and in-repo addons
+   *
+   * Any packageInfos that we find that are marked as not valid are excluded.
+   *
+   * @private
+   * @method discoverAddons
    */
   discoverAddons() {
-    let addonsList = this.addonDiscovery.discoverChildAddons(this);
+    let pkgInfo = this._packageInfoCache.getEntry(this.root);
+
+    let addonPackageList = [];
+
+    // There are cases where an addon can be created in memory and not actually
+    // have a root entry (or have one that is not actually pointing to a directory,
+    // like 'foo' in some of the tests. We don't want to crash, but want to let
+    // things continue even though they're empty.
+    if (pkgInfo) {
+
+      // The 'dependencyPackages' object contains ALL dependencies, not just those
+      // that are addons. We want only the addons.
+      if (pkgInfo.dependencyPackages) {
+        Object.keys(pkgInfo.dependencyPackages).forEach(name => {
+          if (name !== 'ember-cli') {
+            let dependencyInfo = pkgInfo.dependencyPackages[name];
+            if (dependencyInfo.isAddon()) {
+              addonPackageList.push(dependencyInfo);
+            }
+          }
+        });
+      }
 
-    this.addonPackages = this.addonDiscovery.addonPackages(addonsList);
+      if (pkgInfo.inRepoAddons) {
+        pkgInfo.inRepoAddons.forEach(addon => addonPackageList.push(addon));
+      }
+    }
+
+    // Create an AddonInfo for each child addon PackageInfo, so we can
+    // check shouldIncludeChildAddon(), if necessary, and create
+    // this.addonPackages.
+    this.addonPackages = Object.create(null);
+
+    let invalidAddonPackages = [];
+
+    addonPackageList.forEach(packageInfo => {
+      if (packageInfo.valid) {
+        let addonInfo = new AddonInfo(
+          packageInfo.name,
+          packageInfo.realPath,
+          packageInfo.pkg
+        );
+
+        if (!this.shouldIncludeChildAddon || this.shouldIncludeChildAddon(addonInfo)) {
+          this.addonPackages[packageInfo.name] = addonInfo;
+        }
+      } else {
+        invalidAddonPackages.push(packageInfo);
+      }
+    });
+
+    if (invalidAddonPackages.length > 0) {
+      this.ui.writeWarnLine(`For addon at path ${this._packageInfo.realPath}:`);
+      invalidAddonPackages.forEach(packageInfo => {
```

Comment:
is additional iteration necessary here? can we move `this.ui.writeWarnLine` to the `else` branch of the `if` above?

Reply:
I was doing it separately so that we get a single list like the following:

For addon at path XXX:
    Excluding invalid/malformed/missing addon at relative path YYY
    Excluding invalid/malformed/missing addon at relative path ZZZ

I think what you're suggesting results in either no header line about the XXX addon, or duplications of that message. Or am I mistaken?

Reply:
In the above, the 'excluding' lines are supposed to be indented 2 spaces.

Reply:
I see what you mean, makes sense!